Add editor specific replace vars

igorschoester opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Moves some snippet variables to only appear within specific editors. Adds a filter `wpseo_editor_specific_replace_vars` to make this pluggable.

## Relevant technical choices:

Due to time constraints I stopped with writing the tests and created a separate issue for finishing this: https://github.com/Yoast/wordpress-seo/issues/10180

* Editor specific replace vars is an array of page types, each with a list of replace var names.
* Add a filter `wpseo_editor_specific_replace_vars` to make this pluggable.
* Extract the `get_custom_fields` and `get_custom_taxonomies` to their own classes.
* Filters the replacement variables in `SettingsReplacementVariableEditors` so that each editor on the `settings -> search appearance` page receives only the relevant replacement variables.

## Test instructions

This PR can be tested by following these steps:

* Ensure all the snippet editors include the following replacement variables (these are shared):

Replacement variable | Nice name (English)
-- | --
date | Date
title | Title
archive_title | Archive title
sitename | Site title
excerpt | Excerpt
excerpt_only | Excerpt only
tag | Tag
category | Category
primary_category | Primary category
sep | Separator
modified | Modified
name | Name
user_description | User description
page | Page
pagetotal | Pagetotal
pagenumber | Pagenumber
caption | Caption
focuskw | Focus keyword

* Then for the following editors these variables are also there:

Page type (context) | Editor specific replacement variables
-- | --
post | id, pt_single, pt_plural, cf_<custom-field-name>, ct_<custom-tax-name>, ct_desc_<custom-tax-name>, term404
page | id, pt_single, pt_plural, cf_<custom-field-name>, parent_title
custom post type (custom-post-type) | id, pt_single, pt_plural, cf_<custom-field-name>, term404
category | category_description, term_description, term_title, parent_title
tag (post_tag) | tag_description, term_description, term_title
custom taxonomy (term-in-custom-taxonomy) | category_description, term_description, ct_<custom-tax-name>, ct_desc_<custom-tax-name>, parent_title
search | searchphrase


## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes #10000

maartenleenders commented on inc/class-wpseo-custom-taxonomies.php:
```diff
@@ -0,0 +1,69 @@
+	 * @return string The prefixed taxonomy.
+	 */
+	private static function add_custom_taxonomies_description_prefix( $taxonomy ) {
+		return 'ct_desc_' . $taxonomy;
```
Creating these private static functions is for readability's sake, I presume? Fine by me, but I'd argue that the improvement is marginal over the way it was in `class-wpseo-replace-vars` before ;). Still, more readable is more better <3

maartenleenders commented on js/src/components/SettingsReplacementVariableEditors.js:
```diff
@@ -22,30 +24,51 @@ class SettingsReplacementVariableEditors extends React.Component {
 
+	filterEditorSpecificReplaceVars( replacementVariables, pageType ) {
+		const {
+			editor_specific_replace_vars: editorSpecificReplaceVars = {},
```
Wew, learned something new today. Thought this had to be the other way around. Neat! ;)

maartenleenders commented:
CR 👍 
Lots of moving parts, so I hope I didn't miss anything

igorschoester commented:
* Fixed the bug where taxonomies where processed as archives 😞 
* Added `parent_title` to page, category and term in custom taxonomy

afercia commented on admin/class-admin-editor-specific-replace-vars.php:
```diff
@@ -0,0 +1,257 @@
+		'page'                    => array( 'id', 'pt_single', 'pt_plural', 'parent_title' ),
+		'post'                    => array( 'id', 'term404', 'pt_single', 'pt_plural' ),
+		// Custom post type.
+		'custom_post_type'        => array( 'id', 'term404', 'pt_single', 'pt_plural' ),
```
A custom post type could be hierarchical and thus have a parent_title, should this be taken into consideration?

boblinthorst replied:
You're right, I forgot that one in my code review. Sorry @igorschoester

igorschoester replied:
Ah, will add. Could do a check with `is_post_type_hierarchical` too. Seems a little overboard for now though.

afercia commented on admin/class-admin-editor-specific-replace-vars.php:
```diff
@@ -0,0 +1,257 @@
+		'term-in-custom-taxomomy' => array( 'term_title', 'term_description', 'parent_title' ),
+
+		// Settings - archive pages.
+		'custom-post-type_archive' => array(),
```
minor: all the alignment should align with this line `=>`

afercia commented on admin/class-admin-editor-specific-replace-vars.php:
```diff
@@ -0,0 +1,257 @@
+	}
+
+	/**
+	 * Determines the page type of the current term.
```
as a complete newbie to this code, I'm not sure I understand what "page type" means here (same in the following methods). For example: is `post_tag` a "page type"?

igorschoester replied:
Yes to your question. I determined the name with @andizer in our recommended replace vars. It is basically just the name of the key in our array with replace vars. It is supposed to indicate to which context those editor specific replace vars correspond.

afercia commented on admin/class-admin-editor-specific-replace-vars.php:
```diff
@@ -0,0 +1,257 @@
+	/**
+	 * Determines the page type of the current post.
+	 *
+	 * @param WP_Post $post The WordPress global post object.
```
Although seems to me this method is not actually used, for analogy with the one in `WPSEO_Admin_Recommended_Replace_Vars` seems to me this param is not always necessarily the global post, and it could be a retrieved post instance?

igorschoester replied:
This will be needed when implementing for posts and pages. And true, good point, will adapt.

afercia commented on inc/class-wpseo-custom-fields.php:
```diff
@@ -0,0 +1,67 @@
+		 * Filters the number of custom fields to retrieve for the drop-down
+		 * in the Custom Fields meta box.
+		 *
+		 * @since 2.1.0
```
this is the WordPress @since `meta_form()` in `wp-admin/includes/template.php` not sure we should use it?

afercia commented on inc/class-wpseo-custom-fields.php:
```diff
@@ -0,0 +1,67 @@
+			HAVING meta_key NOT LIKE %s
+			ORDER BY meta_key
+			LIMIT %d";
+		$fields        = $wpdb->get_col( $wpdb->prepare( $sql, $wpdb->esc_like( '_' ) . '%', $limit ) );
```
not sure why there's this indentation, please see also lines 41 and 42

afercia commented on js/src/components/SettingsReplacementVariableEditors.js:
```diff
@@ -22,13 +24,28 @@ class SettingsReplacementVariableEditors extends React.Component {
+		const replaceVarNames = [ ...sharedReplaceVars, ...pageTypeSpecificReplaceVars ];
+
+		return replacementVariables.filter( replaceVar => {
+			return replaceVarNames.includes( replaceVar.name );
```
I guess this will error in IE11, as far as I know `includes`  it's not supported. For instance, Gutenberg uses lodash `includes()`.

afercia commented on js/src/components/SettingsReplacementVariableEditors.js:
```diff
@@ -22,13 +24,28 @@ class SettingsReplacementVariableEditors extends React.Component {
+	 * replacevar-title-field-id, data-react-replacevar-metadesc-field-id,
+	 * data-react-replacevar-page-type-recommend and
+	 * data-react-replacevar-page-type-specific attributes to function properly.
 	 * These attributes should point to existing (hidden) inputs in the DOM.
```
I guess this line should now be changed to clarify _which_ attributes should point to hidden fields.

afercia commented on js/src/components/SettingsReplacementVariableEditors.js:
```diff
@@ -60,9 +83,10 @@ class SettingsReplacementVariableEditors extends React.Component {
+	 * react-replacevar-field-id, data-react-replacevar-page-type-recommended
+	 * and data-react-replacevar-page-type-specific attributes to function
+	 * properly. This attribute should point to and existing (hidden) input in
+	 * the DOM.
```
Not clear _which_ attributes should point to hidden fields. Also, "This" or "These"? "point to and existing" should be "point to an existing". I guess it's only about `data-react-replacevar-field-id`.

afercia commented:
CR 🚧 

Admittedly, not my area of expertise 🙂and pretty large to fully follow for me. However, apart from a few minor considerations mostly related to comments/docblocks, I've tried to check the requirements from #10000. Trying to recap:

**%%category_description%% should only show up for categories and custom taxonomies**
it doesn't show up for custom taxonomies; when manually typed, it works

**%%tag_description%% should only show up for tags**
OK, but %%tag%% shows up for all taxonomies (categories, custom non-hierarchical taxonomies) which doesn't make much sense to me

**%%term_description%% and %%term_title%% should only show up for categories, tags, and custom taxonomies**
they show up also for post_format; %%term_description%% doesn't work for them (unsure if there's a way to add descriptions to post formats)

**%%ct_<custom-tax-name>%% and %%ct_desc_<custom-tax-name>%% should only show up for custom taxonomies and posts**
- they show up also for post_format (which is considered "term-in-custom-taxomomy")
- they show up for all custom taxonomies; for example activate jetpack and the Yoast Development Test Helper plugin: type `%j` in the "movie genre" editor and %%ct_jetpack-portfolio-type%% will show up, which doesn't make much sense for the movie genre? Same, the book category will show up for the jetpack project type etc.
- post type "post": type `%j`, `%b`, or `%m`: the jetpack, book, and movies categories show up which doesn't make much sense to me
- post type "book": typing `%b` nothing happens: the book taxonomies should show up
- post type "movie": typing `%m` nothing happens: the movie taxonomies should show up
- post type "jetpack-portfolio": typing `%j` nothing happens: the jetpack-portfolio taxonomies should show up
- post type "jetpack-testimonial": typing `%j` nothing happens: the jetpack-testimonial taxonomies should show up

**%%term404%% should only show up for for posts and custom post types**
- OK but why it shouldn't show up also for pages?

**%%wc_shortdesc%%, %%wc_sku%%, %%wc_brand%% and %%wc_price%% should only be shown for products (activate WooCommerce and WordPress SEO for Woo to see these)**
They appear in the edit product screen in the metabox editor; they do not appear in Search appearance > Content Types > Products

**%%pt_single%% and %%pt_plural%% should only be shown for posts, pages, and custom post types**
- shown also for the Homepage in Search Appearance > General > Homepage, I guess this is OK

**%%cf_<custom-field-name>%% should only show up on posts, pages and custom post types (in the metabox) and on all editors on the Content Types tab (in search appearances)**
- of course, in the various post types metaboxes only the custom field added to that specific post show up 
- instead, in Search Appearance > Content Types: all existing custom fields show up for all the content types so, for example, I can set %%cf_total_sales%% (which is used by WooCommerce) on the standard post or page title or any other cpt, which doesn't make much sense to me

igorschoester commented:
Very nice feedback @afercia ! Thank you 🙇 
I processed a lot of your feedback. Some questions remain I'm afraid.

> **%%category_description%% should only show up for categories and custom taxonomies**
> it doesn't show up for custom taxonomies; when manually typed, it works

Fixed!

> **%%tag_description%% should only show up for tags**
> OK, but %%tag%% shows up for all taxonomies (categories, custom non-hierarchical taxonomies) which doesn't make much sense to me

Now you mention it. It makes sense that `tag` should be a specific replace var too. Would like some confirmation on this. What do you think @jdevalk ?

> **%%term_description%% and %%term_title%% should only show up for categories, tags, and custom taxonomies**
> they show up also for post_format; %%term_description%% doesn't work for them (unsure if there's a way to add descriptions to post formats)

Added `post_format` as `page_type` that does not have any specific replace vars.

> **%%ct_<custom-tax-name>%% and %%ct_desc_<custom-tax-name>%% should only show up for custom taxonomies and posts**
> - they show up also for post_format (which is considered "term-in-custom-taxomomy")

Fixed by the above.

> - they show up for all custom taxonomies; for example activate jetpack and the Yoast Development Test Helper plugin: type `%j` in the "movie genre" editor and %%ct_jetpack-portfolio-type%% will show up, which doesn't make much sense for the movie genre? Same, the book category will show up for the jetpack project type etc.

Do you mean you would want a split in actual custom taxonomies? At this point I have not thought about a way to make a thing like that happen. All the custom taxonomies (`ct_`) are treated the same. As are custom fields (`cf_`).

> - post type "post": type `%j`, `%b`, or `%m`: the jetpack, book, and movies categories show up which doesn't make much sense to me
> - post type "book": typing `%b` nothing happens: the book taxonomies should show up
> - post type "movie": typing `%m` nothing happens: the movie taxonomies should show up
> - post type "jetpack-portfolio": typing `%j` nothing happens: the jetpack-portfolio taxonomies should show up
> - post type "jetpack-testimonial": typing `%j` nothing happens: the jetpack-testimonial taxonomies should show up

I read the line in the issue as only for `custom taxonomies` and `posts`. But reading this I realize this should probably be `custom post types` too. @jdevalk Could you clarify where `ct_` and `ct_desc_` replace vars should be used? 

> **%%term404%% should only show up for for posts and custom post types**
> - OK but why it shouldn't show up also for pages?

@jdevalk Would you mind answering this one?

> **%%wc_shortdesc%%, %%wc_sku%%, %%wc_brand%% and %%wc_price%% should only be shown for products (activate WooCommerce and WordPress SEO for Woo to see these)**
> They appear in the edit product screen in the metabox editor; they do not appear in Search appearance > Content Types > Products

Would have been nice if I had made a solution for this in WordPress SEO for WooCommerce.

> **%%pt_single%% and %%pt_plural%% should only be shown for posts, pages, and custom post types**
> - shown also for the Homepage in Search Appearance > General > Homepage, I guess this is OK

A homepage is also a page. So I removed the exception it had for the recommended replace vars.

> **%%cf_<custom-field-name>%% should only show up on posts, pages and custom post types (in the metabox) and on all editors on the Content Types tab (in search appearances)**
> - of course, in the various post types metaboxes only the custom field added to that specific post show up 

This PR does not change any replacement variables in the metabox.

> - instead, in Search Appearance > Content Types: all existing custom fields show up for all the content types so, for example, I can set %%cf_total_sales%% (which is used by WooCommerce) on the standard post or page title or any other cpt, which doesn't make much sense to me

Same response here as above. I made no distinction for `cf_` replace vars.

IreneStr commented:
Let's get this PR through the process. I've created a new issue for the open questions: https://github.com/Yoast/wordpress-seo/issues/10273

IreneStr commented:
@afercia Thenks for your very thorough review. I've created a new issue for to open questions, in order to keep this PR moving: #10273

boblinthorst commented:
CR: only looking at the commit added after the review by @afercia: 👍

boblinthorst commented:
acceptance: ok 👍
